<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Computational Models for Complex system - The Repressilator</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Computational Models for Complex system</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-site-map" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Site Map</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-site-map">    
        <li>
    <a class="dropdown-item" href="../notebooks/long_range_simulations.html" rel="" target="">
 <span class="dropdown-text">Analysis of long-range ecological models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../notebooks/lattice_simulations.html" rel="" target="">
 <span class="dropdown-text">Analysis of ecological lattice models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../notebooks/repressilator.html" rel="" target="">
 <span class="dropdown-text">Analysis of the Repressilator model</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/iretes/rock-scissors-paper" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#modeling-protein-dynamics-and-gene-regulation" id="toc-modeling-protein-dynamics-and-gene-regulation" class="nav-link active" data-scroll-target="#modeling-protein-dynamics-and-gene-regulation">Modeling protein dynamics and gene regulation</a>
  <ul class="collapse">
  <li><a href="#basic-protein-production-model" id="toc-basic-protein-production-model" class="nav-link" data-scroll-target="#basic-protein-production-model">Basic protein production model</a></li>
  <li><a href="#gene-regulation-and-repressor-binding-dynamics" id="toc-gene-regulation-and-repressor-binding-dynamics" class="nav-link" data-scroll-target="#gene-regulation-and-repressor-binding-dynamics">Gene regulation and repressor binding dynamics</a></li>
  <li><a href="#steady-state-approximation-and-protein-production-rate" id="toc-steady-state-approximation-and-protein-production-rate" class="nav-link" data-scroll-target="#steady-state-approximation-and-protein-production-rate">Steady-State approximation and protein production rate</a></li>
  <li><a href="#ultrasensitivity-and-hill-function" id="toc-ultrasensitivity-and-hill-function" class="nav-link" data-scroll-target="#ultrasensitivity-and-hill-function">Ultrasensitivity and Hill function</a></li>
  </ul></li>
  <li><a href="#description-and-analysis-of-the-repressilator-dynamics" id="toc-description-and-analysis-of-the-repressilator-dynamics" class="nav-link" data-scroll-target="#description-and-analysis-of-the-repressilator-dynamics">Description and analysis of the Repressilator dynamics</a>
  <ul class="collapse">
  <li><a href="#the-repressilator-model" id="toc-the-repressilator-model" class="nav-link" data-scroll-target="#the-repressilator-model">The Repressilator model</a></li>
  <li><a href="#dimensionless-formulation" id="toc-dimensionless-formulation" class="nav-link" data-scroll-target="#dimensionless-formulation">Dimensionless formulation</a></li>
  <li><a href="#oscillatory-behavior-and-stability-analysis" id="toc-oscillatory-behavior-and-stability-analysis" class="nav-link" data-scroll-target="#oscillatory-behavior-and-stability-analysis">Oscillatory behavior and stability analysis</a></li>
  <li><a href="#design-considerations-for-oscillatory-systems" id="toc-design-considerations-for-oscillatory-systems" class="nav-link" data-scroll-target="#design-considerations-for-oscillatory-systems">Design considerations for oscillatory systems</a></li>
  <li><a href="#incorporating-mrna-dynamics-and-leaky-expression" id="toc-incorporating-mrna-dynamics-and-leaky-expression" class="nav-link" data-scroll-target="#incorporating-mrna-dynamics-and-leaky-expression">Incorporating mRNA dynamics and leaky expression</a></li>
  <li><a href="#applications-of-the-repressilator-in-biological-research" id="toc-applications-of-the-repressilator-in-biological-research" class="nav-link" data-scroll-target="#applications-of-the-repressilator-in-biological-research">Applications of the repressilator in biological research</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">The Repressilator</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>This notebook explores an oscillatory network of transcriptional regulators known as Repressilator. First, we will outline the protein production model, taking into account the effects of gene repressors. Then, we will introduce the Repressilator and simulate its dynamics. The following analysis is based on [1].</p>
<p>This oscillatory network was successfully implemented in <em>Escherichia coli</em>, leading to periodic production of green fluorescent protein by the cells [2]. Initially developed as a proof of concept, this system aimed to show that artificial biological constructs could consistently replicate the behavior of electrical circuits. It also paved the way for the development of biological circuits with more practical purposes, such as systems that can count events or activate functions after a fixed period. These advancements hold potential for applications in drug delivery, nano-construction, and other fields. Additionally, the study of oscillatory networks like the Repressilator is inspired by natural biological systems, such as the circadian clocks in organisms that track and anticipate daily environmental changes. This area of research is significant, as evidenced by the <a href="https://www.nobelprize.org/prizes/medicine/2017/summary/">2017 Nobel Prize</a> awarded to Jeffrey Hall, Michael Rosbash, and Michael Young for their discoveries on how internal clocks and biological rhythms govern human life.</p>
<p>Importing the necessary libraries:</p>
<div class="cell" data-execution_count="1">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="im">from</span> scipy.integrate <span class="im">import</span> odeint</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="modeling-protein-dynamics-and-gene-regulation" class="level2">
<h2 class="anchored" data-anchor-id="modeling-protein-dynamics-and-gene-regulation">Modeling protein dynamics and gene regulation</h2>
<section id="basic-protein-production-model" class="level3">
<h3 class="anchored" data-anchor-id="basic-protein-production-model">Basic protein production model</h3>
<p>We assume that proteins are produced at a constant rate <span class="math inline">\(\beta\)</span>, while they are simultaneously removed through active degradation and dilution during cell growth and division. This removal occurs at a rate proportional to the protein concentration, with the proportionality constant <span class="math inline">\(\gamma\)</span>. Consequently, the rate of change in protein concentration <span class="math inline">\(x\)</span> over time is described by the following ordinary differential equation:</p>
<p><span class="math display">\[\frac{\partial x}{\partial t}=\text{production}-(\text{degradation}+\text{dilution})=\beta-\gamma \cdot x\]</span></p>
</section>
<section id="gene-regulation-and-repressor-binding-dynamics" class="level3">
<h3 class="anchored" data-anchor-id="gene-regulation-and-repressor-binding-dynamics">Gene regulation and repressor binding dynamics</h3>
<p>Genes are not continuously “on” but are regulated by cellular mechanisms that adjust their expression levels based on environmental conditions and the cell’s state. Repressors are crucial in this regulation; these proteins bind to specific DNA sequences near a gene’s promoter, influencing gene expression. When a repressor binds, gene expression decreases, and when it unbinds, expression increases.</p>
<p>The binding and unbinding of repressors can be modeled as a reversible chemical reaction:</p>
<p><span class="math display">\[P+R\underset{k-}{\xrightleftharpoons{k_+}} P_{\text{bound}}\]</span></p>
<p>where <span class="math inline">\(P\)</span> is the unbound promoter, <span class="math inline">\(R\)</span> is the repressor, and <span class="math inline">\(P\)</span> bound is the promoter-repressor complex.</p>
<p>The dynamics of this chemical reaction can be expressed as an ordinary differential equation. Although each DNA site within an individual cell is either bound or unbound and is thus best modeled with discrete variables, averaging over a population of cells allows us to use continuous variables to describe the mean occupancy of the site. Let <span class="math inline">\(r\)</span> be the concentration of repressor, <span class="math inline">\(p\)</span> be the concentration of unbound promoter, <span class="math inline">\(p_{\text{bound}}\)</span> be the concentration of promoter bound with a repressor, then the rate of change of the promoter concentration is given by:</p>
<p><span class="math display">\[\frac{\partial p}{\partial t}=-k_+\cdot p \cdot r + k_-\cdot p_{\text{bound}}\]</span></p>
</section>
<section id="steady-state-approximation-and-protein-production-rate" class="level3">
<h3 class="anchored" data-anchor-id="steady-state-approximation-and-protein-production-rate">Steady-State approximation and protein production rate</h3>
<p>Repressor binding and unbinding to the DNA site occur much faster than changes in protein concentrations. Therefore, on the timescale of protein concentration changes, these binding dynamics quickly reach a steady state, where <span class="math inline">\(\partial{p}/\partial{t}\approx 0\)</span> (i.e., despite the ongoing binding and unbinding, their overall effect remains constant over time). Consequently, the binding dynamics can be approximated by:</p>
<p><span class="math display">\[-k_+\cdot p \cdot r + k_-\cdot p_{\text{bound}}=0\]</span></p>
<p>If <span class="math inline">\(p_{\text{tot}}\)</span> is the total concentration of promoters, bound or not, then <span class="math inline">\(p_{\text{tot}}=p+p_{\text{bound}}\)</span>, and we have</p>
<p><span class="math display">\[-k_+\cdot p \cdot r + k_-\cdot (p_{\text{tot}}-p)=0\]</span></p>
<p>which can be rearranged to give the fraction of promoters that are free to allow transcription,</p>
<p><span class="math display">\[\frac{p}{p_{\text{tot}}}=\frac{k_-}{k_-+k_+\cdot r}=\frac{1}{1+\frac{k_+}{k_-}\cdot r}=\frac{1}{1+\frac{r}{K_d}}\]</span></p>
<p>where we have defined the dissociation constant for repressor-target binding</p>
<p><span class="math display">\[K_d=\frac{k_-}{k_+}\]</span></p>
<p>The rate of protein production should be proportional to the probability of the promoter being unbound, therefore</p>
<p><span class="math display">\[\beta(r)=\beta_0\cdot \frac{p}{p_{\text{tot}}}=\frac{\beta_0}{1+\frac{r}{K_d}}\]</span></p>
</section>
<section id="ultrasensitivity-and-hill-function" class="level3">
<h3 class="anchored" data-anchor-id="ultrasensitivity-and-hill-function">Ultrasensitivity and Hill function</h3>
<p>Although the <span class="math inline">\(\beta(r)\)</span> function effectively captures the behavior of a basic transcriptional regulation model, real-world gene regulation and protein-protein interactions often display more switch-like or ultrasensitive characteristics. Ultrasensitivity can originate from various mechanisms, one key source being cooperativity in molecular interactions. For instance, when a protein binds to one DNA site, it can enhance the binding affinity of a second protein at a nearby site. In such cases, an increasing concentration of one component may initially have little effect but eventually lead to a dramatic response.</p>
<p>Hill function offers a method to analyze systems with ultrasensitive responses. A repressive Hill function is given by</p>
<p><span class="math display">\[f_{\text{rep}}(x)=\frac{1}{1+(\frac{x}{k})^n}\]</span></p>
<p>here, the Hill coefficient, <span class="math inline">\(n\)</span>, describes the degree of ultrasensitivity. For <span class="math inline">\(n=1\)</span>, this reduces to the simple repressive function discussed earlier. When <span class="math inline">\(n&gt;1\)</span>, the response becomes increasingly sharper and more ultrasensitive, approaching a perfect step function as <span class="math inline">\(n\)</span> approaches infinity, as shown in the plot below.</p>
<div class="cell" data-execution_count="2">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1"></a>n_vals <span class="op">=</span> [<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">25</span>, np.inf]</span>
<span id="cb2-2"><a href="#cb2-2"></a>x <span class="op">=</span> np.concatenate((np.linspace(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">200</span>), np.linspace(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">200</span>)))</span>
<span id="cb2-3"><a href="#cb2-3"></a>f_rep <span class="op">=</span> []</span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="cf">for</span> n <span class="kw">in</span> n_vals:</span>
<span id="cb2-5"><a href="#cb2-5"></a>    <span class="cf">if</span> n <span class="op">==</span> np.inf:</span>
<span id="cb2-6"><a href="#cb2-6"></a>        f_rep.append(np.concatenate((np.ones(<span class="dv">200</span>), np.zeros(<span class="dv">200</span>))))</span>
<span id="cb2-7"><a href="#cb2-7"></a>    <span class="cf">else</span>:</span>
<span id="cb2-8"><a href="#cb2-8"></a>        f_rep.append(<span class="dv">1</span> <span class="op">/</span> (<span class="dv">1</span> <span class="op">+</span> x <span class="op">**</span> n))</span>
<span id="cb2-9"><a href="#cb2-9"></a>df <span class="op">=</span> pd.DataFrame(np.array(f_rep).T, columns<span class="op">=</span>[<span class="ss">f"n=</span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> n <span class="kw">in</span> n_vals])</span>
<span id="cb2-10"><a href="#cb2-10"></a>df[<span class="st">'x'</span>] <span class="op">=</span> x</span>
<span id="cb2-11"><a href="#cb2-11"></a>df.plot(</span>
<span id="cb2-12"><a href="#cb2-12"></a>    x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=</span>[<span class="ss">f"n=</span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> n <span class="kw">in</span> n_vals],</span>
<span id="cb2-13"><a href="#cb2-13"></a>    ylabel<span class="op">=</span><span class="st">"Repressing Hill function"</span>, xlabel<span class="op">=</span><span class="st">"$x/k$"</span>,</span>
<span id="cb2-14"><a href="#cb2-14"></a>    colormap<span class="op">=</span><span class="st">"Reds"</span>, linewidth<span class="op">=</span><span class="dv">3</span></span>
<span id="cb2-15"><a href="#cb2-15"></a>)</span>
<span id="cb2-16"><a href="#cb2-16"></a>plt.grid()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="repressilator_files/figure-html/cell-3-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The production rate of gene products under a repressor with ultrasensitivity, modeled by the Hill function, is then</p>
<p><span class="math display">\[\beta(r)=\beta_0\cdot f_{\text{rep}(r)}=\beta_0 \frac{1}{1+(\frac{r}{k})^n}\]</span></p>
<p>To simplify notation, we will omit the subscript <span class="math inline">\(0\)</span> from <span class="math inline">\(\beta\)</span> in the following discussions.</p>
</section>
</section>
<section id="description-and-analysis-of-the-repressilator-dynamics" class="level2">
<h2 class="anchored" data-anchor-id="description-and-analysis-of-the-repressilator-dynamics">Description and analysis of the Repressilator dynamics</h2>
<section id="the-repressilator-model" class="level3">
<h3 class="anchored" data-anchor-id="the-repressilator-model">The Repressilator model</h3>
<p>The Repressilator is a genetic regulatory network consisting of a feedback loop with at least three genes, each expressing a protein that represses the next gene in the loop. The dynamic behavior of the Repressilator mirrors the competitive interactions of the rock-paper-scissors game, as each component cyclically inhibits the next.</p>
<p>The system of differential equations describing the Repressilator is given by</p>
<span class="math display">\[\begin{cases}
\frac{\partial x_1}{\partial t}=\frac{\beta}{1+(\frac{x_3}{k})^n}-\gamma x_1\\
\frac{\partial x_2}{\partial t}=\frac{\beta}{1+(\frac{x_1}{k})^n}-\gamma x_2\\
\frac{\partial x_3}{\partial t}=\frac{\beta}{1+(\frac{x_2}{k})^n}-\gamma x_3\\
\end{cases}\]</span>
<p>or in a more compact form <span class="math display">\[\frac{\partial x_i}{\partial t}=\frac{\beta}{1+(\frac{x_j}{k})^n}-\gamma x_i\]</span> with <span class="math inline">\((i,j)\)</span> pairs <span class="math inline">\((1,3),(2,1),(3,2)\)</span>.</p>
<p>For simplicity we assumed perfect symmetry among the species, with each one having identical biochemical properties.</p>
</section>
<section id="dimensionless-formulation" class="level3">
<h3 class="anchored" data-anchor-id="dimensionless-formulation">Dimensionless formulation</h3>
<p>To simplify the equations, we introduce a dimensionless time variable <span class="math inline">\(\tau\)</span> by rescaling the original time <span class="math inline">\(t\)</span> using <span class="math inline">\(\gamma\)</span>. The new dimensionless time variable <span class="math inline">\(\tau\)</span> is defined as: <span class="math display">\[\tau=\gamma t\]</span> This means that <span class="math inline">\(\tau\)</span> is a measure of time in units of <span class="math inline">\(\gamma^{-1}\)</span>.</p>
<p>The original time derivative <span class="math inline">\(\frac{\partial}{\partial t}\)</span> can be expressed in terms of <span class="math inline">\(\tau\)</span> as: <span class="math display">\[\frac{\partial}{\partial t}=\frac{\partial \tau}{\partial t}\frac{\partial}{\partial \tau}=\gamma \frac{\partial}{\partial \tau}\]</span> since <span class="math inline">\(\frac{\partial \tau}{\partial t}=\gamma\)</span>.</p>
<p>We now define dimensionless concentrations <span class="math inline">\(x_i'\)</span> as: <span class="math display">\[x_i'=\frac{x_i}{k}\]</span> thus <span class="math display">\[x_i=k x_i'\]</span></p>
<p>Substituting the above definitions into the original equations, we get:</p>
<p><span class="math display">\[\frac{\partial x_i}{\partial t}=k \gamma \frac{\partial x_i'}{\partial \tau}\]</span> And since <span class="math display">\[\frac{\partial x_i}{\partial t}= \frac{\beta}{1+(x_j')^n}-\gamma k x_i'\]</span> we have that <span class="math display">\[\frac{\partial x_i'}{\partial \tau}=\frac{\frac{\beta}{k\gamma}}{1+(x_j')^n}- x_i'=\frac{\beta'}{1+(x_j')^n}- x_i'\]</span> where we set <span class="math inline">\(\beta'=\frac{\beta}{k \gamma}\)</span></p>
<p>To simplify the notation in the following we will use <span class="math inline">\(x_i\)</span> instead of <span class="math inline">\(x_i'\)</span>, <span class="math inline">\(\tau\)</span> instead of <span class="math inline">\(t\)</span> and <span class="math inline">\(\beta\)</span> instead of <span class="math inline">\(\beta'\)</span>.</p>
</section>
<section id="oscillatory-behavior-and-stability-analysis" class="level3">
<h3 class="anchored" data-anchor-id="oscillatory-behavior-and-stability-analysis">Oscillatory behavior and stability analysis</h3>
<p>The system’s dynamics might result in sustained oscillation, where repressor proteins alternate in concentration: if repressor 1 is high, it leads to low levels of repressor 2, which allows repressor 3 to increase and suppress repressor 1, continuing the cycle. Alternatively, the system might also stabilize at a steady state, where all three repressors reach intermediate levels, maintaining a balance without oscillations. To effectively design clock circuits, it is essential to ensure that the system exhibits periodic oscillations that persist over time without damping out and can recover from temporary perturbations caused by environmental fluctuations or internal noise.</p>
<p>The system’s actual behavior is influenced by its parameters, specifically the Hill coefficient <span class="math inline">\(n\)</span> and the protein production rate <span class="math inline">\(\beta\)</span>. By integrating the differential equations, we can track the concentrations of three proteins over time for various values of <span class="math inline">\(n\)</span> and <span class="math inline">\(\beta\)</span>, allowing us to observe how these parameters affect the system’s dynamics.</p>
<p>Defining a function to solve the system of differential equations:</p>
<div class="cell" data-execution_count="3">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">def</span> repressilator_ODE_solver(x0, t, beta, n):</span>
<span id="cb3-2"><a href="#cb3-2"></a>    <span class="co">"""</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">    Solve the initial value problem for the Repressilator ODE system.</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">    """</span></span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a>    <span class="kw">def</span> repressilator(x, t, beta, n):</span>
<span id="cb3-7"><a href="#cb3-7"></a>        x1, x2, x3 <span class="op">=</span> x</span>
<span id="cb3-8"><a href="#cb3-8"></a></span>
<span id="cb3-9"><a href="#cb3-9"></a>        d_x1_dt <span class="op">=</span> beta <span class="op">/</span> (<span class="dv">1</span> <span class="op">+</span> x3 <span class="op">**</span> n) <span class="op">-</span> x1</span>
<span id="cb3-10"><a href="#cb3-10"></a>        d_x2_dt <span class="op">=</span> beta <span class="op">/</span> (<span class="dv">1</span> <span class="op">+</span> x1 <span class="op">**</span> n) <span class="op">-</span> x2</span>
<span id="cb3-11"><a href="#cb3-11"></a>        d_x3_dt <span class="op">=</span> beta <span class="op">/</span> (<span class="dv">1</span> <span class="op">+</span> x2 <span class="op">**</span> n) <span class="op">-</span> x3</span>
<span id="cb3-12"><a href="#cb3-12"></a></span>
<span id="cb3-13"><a href="#cb3-13"></a>        <span class="cf">return</span> [d_x1_dt, d_x2_dt, d_x3_dt]</span>
<span id="cb3-14"><a href="#cb3-14"></a></span>
<span id="cb3-15"><a href="#cb3-15"></a>    y <span class="op">=</span> odeint(repressilator, x0, t, args<span class="op">=</span>(beta, n))</span>
<span id="cb3-16"><a href="#cb3-16"></a>    <span class="cf">return</span> y.T</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Solving the system for different values of <span class="math inline">\(n\)</span> and <span class="math inline">\(\beta\)</span>, with initial protein concentrations set to <span class="math inline">\(1\)</span>, <span class="math inline">\(1\)</span>, and <span class="math inline">\(1.1\)</span> for <span class="math inline">\(x_1\)</span>, <span class="math inline">\(x_2\)</span>, and <span class="math inline">\(x_3\)</span>, respectively (we avoided setting <span class="math inline">\(x_1 = x_2 = x_3\)</span> to prevent the dynamics from collapsing into a single equation, which would obscure differences in the proteins’ behavior over time):</p>
<div class="cell" data-execution_count="4">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1"></a>params <span class="op">=</span> {</span>
<span id="cb4-2"><a href="#cb4-2"></a>    <span class="st">'betas'</span>: [<span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">100</span>],</span>
<span id="cb4-3"><a href="#cb4-3"></a>    <span class="st">'ns'</span>: [<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">5</span>]</span>
<span id="cb4-4"><a href="#cb4-4"></a>}</span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a>t <span class="op">=</span> np.linspace(<span class="dv">0</span>, <span class="dv">40</span>, <span class="dv">1000</span>)</span>
<span id="cb4-7"><a href="#cb4-7"></a></span>
<span id="cb4-8"><a href="#cb4-8"></a>df <span class="op">=</span> pd.DataFrame()</span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="cf">for</span> beta, n <span class="kw">in</span> <span class="bu">zip</span>(params[<span class="st">'betas'</span>], params[<span class="st">'ns'</span>]):</span>
<span id="cb4-10"><a href="#cb4-10"></a>    y <span class="op">=</span> repressilator_ODE_solver(x0<span class="op">=</span>np.array([<span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">1.1</span>]), t<span class="op">=</span>t, beta<span class="op">=</span>beta, n<span class="op">=</span>n)</span>
<span id="cb4-11"><a href="#cb4-11"></a>    df_temp <span class="op">=</span> pd.DataFrame(y.T, columns<span class="op">=</span>[<span class="st">"$x_1$"</span>, <span class="st">"$x_2$"</span>, <span class="st">"$x_3$"</span>])</span>
<span id="cb4-12"><a href="#cb4-12"></a>    df_temp[<span class="st">'$</span><span class="ch">\\</span><span class="st">beta$'</span>] <span class="op">=</span> beta</span>
<span id="cb4-13"><a href="#cb4-13"></a>    df_temp[<span class="st">'$n$'</span>] <span class="op">=</span> n</span>
<span id="cb4-14"><a href="#cb4-14"></a>    df_temp[<span class="st">'t'</span>] <span class="op">=</span> t</span>
<span id="cb4-15"><a href="#cb4-15"></a>    df <span class="op">=</span> pd.concat([df, df_temp])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Plotting the results:</p>
<div class="cell" data-execution_count="5">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">4</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">16</span>))</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="cf">for</span> i, (beta, n) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(params[<span class="st">'betas'</span>], params[<span class="st">'ns'</span>])):</span>
<span id="cb5-3"><a href="#cb5-3"></a>    df_temp <span class="op">=</span> df[(df[<span class="st">'$</span><span class="ch">\\</span><span class="st">beta$'</span>] <span class="op">==</span> beta) <span class="op">&amp;</span> (df[<span class="st">'$n$'</span>] <span class="op">==</span> n)]</span>
<span id="cb5-4"><a href="#cb5-4"></a>    df_temp.plot(</span>
<span id="cb5-5"><a href="#cb5-5"></a>        x<span class="op">=</span><span class="st">'t'</span>, y<span class="op">=</span>[<span class="st">"$x_1$"</span>, <span class="st">"$x_2$"</span>, <span class="st">"$x_3$"</span>],</span>
<span id="cb5-6"><a href="#cb5-6"></a>        xlabel<span class="op">=</span><span class="st">"t"</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>        title<span class="op">=</span><span class="ss">f"(</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">)</span><span class="ch">\n</span><span class="ss">$</span><span class="ch">\\</span><span class="ss">beta=</span><span class="sc">{</span>beta<span class="sc">}</span><span class="ss">$, $n=</span><span class="sc">{</span>n<span class="sc">}</span><span class="ss">$"</span>,</span>
<span id="cb5-8"><a href="#cb5-8"></a>        ax<span class="op">=</span>axs[<span class="bu">int</span>(i<span class="op">/</span><span class="dv">3</span>), i<span class="op">%</span><span class="dv">3</span>]</span>
<span id="cb5-9"><a href="#cb5-9"></a>    )</span>
<span id="cb5-10"><a href="#cb5-10"></a>    axs[<span class="bu">int</span>(i<span class="op">/</span><span class="dv">3</span>), i<span class="op">%</span><span class="dv">3</span>].grid()</span>
<span id="cb5-11"><a href="#cb5-11"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="repressilator_files/figure-html/cell-6-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The parameter combinations used in plots <span class="math inline">\((1), (2), (3), (4)\)</span>, and <span class="math inline">\((7)\)</span> lead the system to stabilize at a fixed point, without exhibiting oscillatory behavior.</p>
</section>
<section id="design-considerations-for-oscillatory-systems" class="level3">
<h3 class="anchored" data-anchor-id="design-considerations-for-oscillatory-systems">Design considerations for oscillatory systems</h3>
<p>Through linear stability analysis, it can be proven that the fixed point is unstable (the system exhibits sustained oscillations) for <span class="math inline">\(n&gt;2\)</span> and <span class="math inline">\(\beta &gt; \frac{n}{2}(\frac{n}{2}-1)^{-\frac{n+1}{n}}\)</span>.</p>
<p>We can display the regions of stability and instability in the <span class="math inline">\(n\)</span>-<span class="math inline">\(\beta\)</span> plane, marking the parameter values used in the previous simulations:</p>
<div class="cell" data-execution_count="6">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1"></a>n <span class="op">=</span> np.linspace(<span class="fl">2.01</span>, <span class="dv">5</span>, <span class="dv">200</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a>beta <span class="op">=</span> n <span class="op">/</span> <span class="dv">2</span> <span class="op">*</span> (n <span class="op">/</span> <span class="dv">2</span> <span class="op">-</span> <span class="dv">1</span>) <span class="op">**</span> (<span class="op">-</span>((n <span class="op">+</span> <span class="dv">1</span>)<span class="op">/</span>n))</span>
<span id="cb6-3"><a href="#cb6-3"></a>plt.figure()</span>
<span id="cb6-4"><a href="#cb6-4"></a>plt.semilogy(n, beta, linewidth<span class="op">=</span><span class="dv">3</span>, c<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb6-5"><a href="#cb6-5"></a>plt.fill_between(n, beta, <span class="dv">3000</span>, color<span class="op">=</span><span class="st">"lightgray"</span>, alpha<span class="op">=</span><span class="fl">0.7</span>, label<span class="op">=</span><span class="st">"unstable region</span><span class="ch">\n</span><span class="st">(sustained oscillations)"</span>)</span>
<span id="cb6-6"><a href="#cb6-6"></a>plt.xlabel(<span class="st">"$n$"</span>)</span>
<span id="cb6-7"><a href="#cb6-7"></a>plt.ylabel(<span class="st">"$</span><span class="ch">\\</span><span class="st">beta$"</span>)</span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="cf">for</span> i, (beta, n) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(params[<span class="st">'betas'</span>], params[<span class="st">'ns'</span>])):</span>
<span id="cb6-9"><a href="#cb6-9"></a>    plt.scatter(n, beta, c<span class="op">=</span><span class="st">"black"</span>, zorder<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb6-10"><a href="#cb6-10"></a>    plt.text(n<span class="op">-</span><span class="fl">0.02</span>, beta, <span class="ss">f"(</span><span class="sc">{</span>i<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">)"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, ha<span class="op">=</span><span class="st">'right'</span>)</span>
<span id="cb6-11"><a href="#cb6-11"></a>plt.legend(loc<span class="op">=</span><span class="st">'upper right'</span>)</span>
<span id="cb6-12"><a href="#cb6-12"></a>plt.grid(<span class="va">True</span>, zorder<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<p><img src="repressilator_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The plot indicates that, from a design perspective, maximizing both <span class="math inline">\(\beta\)</span> and <span class="math inline">\(n\)</span> is crucial for ensuring sustained oscillations in the system. This requires employing strong promoters and ultrasensitive repression curves to achieve the desired oscillatory behavior.</p>
</section>
<section id="incorporating-mrna-dynamics-and-leaky-expression" class="level3">
<h3 class="anchored" data-anchor-id="incorporating-mrna-dynamics-and-leaky-expression">Incorporating mRNA dynamics and leaky expression</h3>
<p>While this model focuses exclusively on protein dynamics, it could be enhanced by incorporating mRNA dynamics and considering ‘leaky expression’—the fact that gene repression never completely reduces expression to zero, resulting in a persistent ‘basal’ level. These enhancements provide valuable insights: (1) leaky transcription can diminish oscillations when the residual protein production is enough to suppress the next repressor’s expression, and (2) comparable mRNA and protein decay rates promote oscillations, allowing them to occur with reduced ultrasensitivity (i.e., at lower values of <span class="math inline">\(n\)</span>).</p>
</section>
<section id="applications-of-the-repressilator-in-biological-research" class="level3">
<h3 class="anchored" data-anchor-id="applications-of-the-repressilator-in-biological-research">Applications of the repressilator in biological research</h3>
<p>Riglar et al.&nbsp;[3] used the Repressilator phase as a tool to measure bacterial growth in the mammalian gut, an environment that is difficult to observe directly. By employing the Repressilator as a growth recorder, they were able to track cell growth, revealing how it varies under different conditions, such as inflammation or antibiotic treatment. They discovered that growth variability increased under certain conditions and differed across various spatial niches within the gut. This approach provided insights into bacterial growth patterns that would otherwise be difficult to observe.</p>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<ul>
<li>[1] Elowitz M., Bois J. and Marken J. (2022). <a href="https://biocircuits.github.io/chapters/09_repressilator.html">“Biological Circuits Design”</a>. California Institute of Technology.</li>
<li>[2] Elowitz M. and Leibler S. “A synthetic oscillatory network of transcriptional regulators.” Nature 403.6767 (2000): 335-338.</li>
<li>[3] Riglar, David T., et al.&nbsp;“Bacterial variability in the mammalian gut captured by a single-cell synthetic oscillator.” Nature communications 10.1 (2019): 4665.</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">© 2024 Irene Testa</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>